src/transactions/TransactionFrame.cpp

```diff
@@ -681,8 +684,14 @@ TransactionFrame::apply(Application& app, AbstractLedgerTxn& ltx,
         {
             processSeqNum(ltxTx);
         }
-        auto signaturesValid = cv >= (ValidationType::kInvalidPostAuth) &&
-                               processSignatures(signatureChecker, ltxTx);
+        auto signaturesValid =
+            cv >= (ValidationType::kMaybeValidCheckSignatures) &&
+            processSignatures(signatureChecker, ltxTx);
```

Comment:
As I added a comment on the scope around `LedgerTxn ltx2(ltx)`, I realized that:
* either this scope is not needed (pretty much sure it is)
* or we have a potential bug with the new `removeOneTimeSignerFromAllSourceAccounts` that loads all source accounts - it should only load & modify the ones that get changed

Reply:
> I don't understand the name kMaybeValidCheckSignatures: it's invalid for sure as it failed with txINSUFFICIENT_BALANCE.

> As we make this change, I think it's better to "fast fail" with the proper error code in protocol 13. This way the transaction will fail with txINSUFFICIENT_BALANCE instead of (potentially) some other error like txBAD_AUTH_EXTRA.

Yes, you're absolutely right about this. When I wrote that comment I didn't notice that we check the balance then check the signatures regardless of the result of the balance check. In an ideal world, we would never have even checked signatures in this case. Let's make this change in protocol 13. Let's also rename `kFullyValid` to `kMaybeValid` as you suggest.

Reply:
Also in the code @MonsieurNicolas suggested, I don't believe that `LedgerTxn ltx(ltxOuter);` (line 8) is required. However, the scope with `LedgerTxn ltx2(ltx);` is required (but let's just call it `ltx` now).

You are correct that `removeOneTimeSignerFromAllSourceAccounts` is currently wrong, it should only load and modify the ones that get changed.

Reply:
I modified `removeOneTimeSignerFromAllSourceAccounts` to first look for the signer through `loadAccountWithoutRecord`. 

Something to scrutinize here is that I use the iterator derived from the `loadAccountWithoutRecord` account signers to erase from the `loadAccount` account signers to avoid a second search. This should be fine since the `LedgerEntry` from the first load should be cached, but I'm wondering if it might be safer and cleaner to just do the second search if we get a match on the signer (which isn't a common scenario). @jonjove 

I also added a check in `TxTests.cpp` that verifies any delta from an early failure post V13 must have a signature removed.

Reply:
This isn't safe. Every time a new `LedgerEntry` is loaded it produces a copy. So the vector of signers for the mutable `LedgerTxnEntry` will be a copy of your initial vector from the `ConstLedgerTxnEntry`; the initial vector may even have been destructed.

That being said, you don't actually need to search again. The signers are sorted which guarantees they won't change order, so you can calculate the index of the relevant signer. Then you can just index into the new vector directly (I would recommend adding an assert here to check that you actually found the right signer).

Reply:
Fixed